src/main/kotlin/org/jetbrains/research/kotlin/mpp/inference/operators/layer/reccurent/lstm/LSTMLayer.kt

```diff
@@ -0,0 +1,126 @@
+                                    prevState: State<T>): GatesData<T> {
+                val gates = inputMatrix.dot(weights.transpose()) + prevState.output.dot(recWeights.transpose())
+                val gatesList = gates.splitByIndex(4, 1)
+                return GatesData(gatesList[0], gatesList[1], gatesList[2], gatesList[3])
+            }
+        }
+    }
+
+    data class State<T : Number>(val output: Tensor<T>, val cellGate: Tensor<T>) {
+        companion object {
+            @Suppress("UNCHECKED_CAST")
+            fun <T : Number> initialize(batchSize: Int, hiddenSize: Int, type: TensorProto.DataType): State<T> {
+                val stateSpace = resolveSpaceWithKClass(type.resolveKClass(), intArrayOf(batchSize, hiddenSize)) as TensorRing<T>
+                return State(Tensor(null, stateSpace.zero, type, stateSpace), Tensor(null, stateSpace.zero, type, stateSpace))
+            }
+        }
+    }
+
+    data class BiasesData<T : Number>(val inputGateBiases: Tensor<T>,
+                                      val outputGateBiases: Tensor<T>,
+                                      val forgetGateBiases: Tensor<T>,
+                                      val cellGateBiases: Tensor<T>) {
+        companion object {
+            fun <T : Number> create(biases: Tensor<T>, hiddenSize: Int, batchSize: Int) : Pair<BiasesData<T>, BiasesData<T>> {
+                val shape = intArrayOf(hiddenSize, batchSize)
+                val blockSize = hiddenSize * batchSize
+                val newStrides = SpaceStrides(shape)
+                @Suppress("UNCHECKED_CAST")
+                val newSpace = resolveSpaceWithKClass(biases.type!!.resolveKClass(), shape) as TensorRing<T>
+                val parsedBiases = List(8) { index ->
+                    val newBuffer = VirtualBuffer(blockSize) { i ->
+                        val indices = newStrides.index(i)
+                        val rowNum = indices[0]
+                        biases.data.buffer[hiddenSize * index + rowNum]
+                    }
+                    val newStructure = BufferNDStructure(newStrides, newBuffer)
+                    Tensor(null, newStructure, biases.type, newSpace)
+                }
+                val weightsBiasesData = BiasesData(parsedBiases[0], parsedBiases[1], parsedBiases[2], parsedBiases[3])
+                val recursiveWeightsBiasesData = BiasesData(parsedBiases[4], parsedBiases[5], parsedBiases[6], parsedBiases[7])
+                return Pair(weightsBiasesData, recursiveWeightsBiasesData)
+            }
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun <T : Number> List<Tensor<T>>.toOutput(): Collection<Tensor<T>> {
+        val newShape = intArrayOf(this.size, 1, this.first().data.shape[0], this.first().data.shape[1])
+        val newStrides = SpaceStrides(newShape)
+        val newData = VirtualBuffer(newStrides.linearSize) { i ->
+            val indices = newStrides.index(i)
+            val inputNum = indices[0]
```

Comment:
BTW, is it guaranteed that there is only 4 values? If yes -- we can do `val (inputNum, _ , rowNum, colNum) = indices`